flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataset/DataSetJoinRule.scala

```diff
@@ -39,6 +46,10 @@ class DataSetJoinRule
     val convLeft: RelNode = RelOptRule.convert(join.getInput(0), DataSetConvention.INSTANCE)
     val convRight: RelNode = RelOptRule.convert(join.getInput(1), DataSetConvention.INSTANCE)
 
+    val joinKeys = getJoinKeys(join)
+
+    // There would be a FlinkProject after FlinkJoin to handle the output fields afterward join,
+    // so we do not need JoinFunction here by now.
```

Comment:
We need the `JoinFunction` to evaluate all non-equality join predicates. For instance if we have a join predicate such as `t1.a == t2.a AND t1.b < t2.b`, we would evaluate the `t1.a == t2.a` part by the joinKeys of the JoinOperator and the `t1.b < t2.b` as part of the `JoinFunction`.